packages/cozy-harvest-lib/src/models/ConnectionFlow.js

```diff
@@ -0,0 +1,529 @@
+ * @param  {CozyClient} options.client - A cozy client
+ * @param  {io.cozy.konnector} options.konnector - Konnector to which the account is linked
+ * @param  {KonnectorPolicy} options.konnectorPolicy - Controls if auth is saved in io.cozy.accounts
+ * and if auth is saved into the vault
+ * @param  {function} options.saveAccount
+ * @param  {object} options.userCredentials
+ */
+export const createOrUpdateAccount = async ({
+  account,
+  cipher,
+  client,
+  flow,
+  konnector,
+  konnectorPolicy,
+  userCredentials
+}) => {
+  assert(flow, 'No flow')
+  assert(userCredentials, 'No user credentials')
+  const isUpdate = !!account
+  const { onAccountCreation, saveInVault } = konnectorPolicy
+
+  let accountToSave = clone(account) || {}
+
+  accountToSave = accounts.resetState(accountToSave)
+
+  accountToSave = accounts.setSessionResetIfNecessary(
+    accountToSave,
+    userCredentials
+  )
+
+  accountToSave = isUpdate
+    ? accounts.mergeAuth(accountToSave, userCredentials)
+    : accounts.build(konnector, userCredentials)
+
+  if (onAccountCreation) {
+    logger.debug(
+      `Using konnector policy "${konnectorPolicy.name}"'s custom  account creation`
+    )
+    accountToSave = await onAccountCreation({
+      client,
+      flow,
+      account: accountToSave,
+      konnector
+    })
+  }
+
+  if (cipher && saveInVault) {
+    accountToSave = accounts.setVaultCipherRelationship(
+      accountToSave,
+      cipher.id
+    )
+  } else {
+    logger.warn(
+      'No cipher passed when creating/updating account, account will not be linked to cipher'
+    )
+  }
+
+  return await saveAccount(client, konnector, accountToSave)
+}
+
+/**
+ * Event hub to launch and follow a konnector job.
+ *
+ * - Creates the job for a trigger
+ * - Listens to changes on
+ *   - the created job
+ *   - the account linked to the trigger
+ *
+ * Transforms low-level events happening on the documents to higher-level events
+ * closer to business logic.
+ *
+ * This should be the go to source of truth for the state of a Konnector Job.
+ */
+export class ConnectionFlow {
+  constructor(client, trigger) {
+    this.client = client
+    this.trigger = trigger
+    this.unsubscribeAllRealtime = null
+
+    // Bind methods used as callbacks
+    this.getTwoFACodeProvider = this.getTwoFACodeProvider.bind(this)
+    this.getKonnectorSlug = this.getKonnectorSlug.bind(this)
+    this.handleAccountUpdated = this.handleAccountUpdated.bind(this)
+    this.handleJobUpdated = this.handleJobUpdated.bind(this)
+    this.handleTwoFA = this.handleTwoFA.bind(this)
+    this.launch = this.launch.bind(this)
+    this.sendTwoFACode = this.sendTwoFACode.bind(this)
+    this.unwatch = this.unwatch.bind(this)
+
+    this.state = {
+      status: IDLE
+    }
+
+    // Stores data necessary for custom connection flow, for example
+    // a budget-insight temporary token
+    this.data = {}
+
+    this.twoFAWaiters = this.twoFAWaiters || []
```

Comment:
How could `this.twoFAWaiters` be already defined?

Reply:
If the function has already been called once.